### PR TITLE
Add dimmer

### PR DESCRIPTION
**Pull request by JELoohuis**
- **Added support for the dimmer category of devices.**
  The following capabilities were implemented and tested:
  - onoff (switch_led_1, switch_led_2)
  - dim (bright_value_1, bright_value_2)

  The onoff is handled the same as in sockets, with 1 channel devices using the default capability,
  and 2 channel devices using it to turn both channels on/off, with sub-capabilities for the separate channels.

  In 2 channel devices the default dim capability is removed in favor of just sub-capabilities.

  The following settings were implemented and tested:
  - Minimum brightness (brightness_min_1, brightness_min_2)
  - Maximum brightness (brightness_max_1, brightness_max_2)
  - Bulb type (led_type_1, led_type_2)

  The work_mode instruction could not be tested and was not added.

### CODE REVIEW

File: drivers/dimmer/device.js [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+        });
+      }
+
+      if (typeof brightnessStatus === "number") {
+        const scaleMin =
+          this.getSetting(tuyaBrightnessMin) * TUYA_PERCENTAGE_SCALING;
+        const scaleMax =
+          this.getSetting(tuyaBrightnessMax) * TUYA_PERCENTAGE_SCALING;
+        const scaledValue =
+          (brightnessStatus - scaleMin) / (scaleMax - scaleMin);
```
Comment (bobvandevijver) [inline drivers/dimmer/device.js:98]:
```suggestion
        const scaleMin = this.getSetting(tuyaBrightnessMin) * TUYA_PERCENTAGE_SCALING;
        const scaleMax = this.getSetting(tuyaBrightnessMax) * TUYA_PERCENTAGE_SCALING;
        const scaledValue = (brightnessStatus - scaleMin) / (scaleMax - scaleMin);
```

File: drivers/dimmer/device.js [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+          );
+          triggerCard
+            .trigger(
+              this,
+              {
+                value: Math.round(scaledValue * 100) / 100, // Round to 2 decimals
+              },
+              {},
+            )
+            .catch(this.error);
```
Comment (bobvandevijver) [inline drivers/dimmer/device.js:112]:
```suggestion
          triggerCard.trigger(this, {
              value: Math.round(scaledValue * 100) / 100, // Round to 2 decimals
          }).catch(this.error);
```

File: drivers/dimmer/device.js [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    const unsupportedValues = [];
+
+    for (const changedKey of changedKeys) {
+      const newValue = newSettings[changedKey];
+      await this.sendCommand({
+        code: changedKey,
+        value:
+          typeof newValue === "number"
+            ? newValue * TUYA_PERCENTAGE_SCALING
+            : newValue,
```
Comment (bobvandevijver) [inline drivers/dimmer/device.js:136]:
```suggestion
        value: typeof newValue === "number"
            ? newValue * TUYA_PERCENTAGE_SCALING
            : newValue,
```

File: drivers/dimmer/device.js [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+      });
+    }
+
+    // Report back which capabilities and values are unsupported,
+    // since we cannot programmatically remove settings.
+    const messages = [];
+
+    if (unsupportedSettings.length > 0) {
+      let unsupportedSettingsMessage =
+        this.homey.__("settings_unsupported") + " ";
```
Comment (bobvandevijver) [inline drivers/dimmer/device.js:154]:
```suggestion
      let unsupportedSettingsMessage = this.homey.__("settings_unsupported") + " ";
```

File: drivers/dimmer/device.js [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+      const mappedSettingNames = unsupportedSettings.map(
+        (settingKey) => DIMMER_SETTING_LABELS[settingKey],
+      );
+      unsupportedSettingsMessage += mappedSettingNames.join(", ");
+      messages.push(unsupportedSettingsMessage);
+    }
+    if (unsupportedValues.length > 0) {
+      let unsupportedValuesMessage =
+        this.homey.__("setting_values_unsupported") + " ";
```
Comment (bobvandevijver) [inline drivers/dimmer/device.js:163]:
```suggestion
      let unsupportedValuesMessage = this.homey.__("setting_values_unsupported") + " ";
```

File: drivers/dimmer/device.js [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+  }
+
+  async singleDim(value, tuyaCapability) {
+    const subSwitch = tuyaCapability.at(tuyaCapability.length - 1);
+    const scaleMin =
+      this.getSetting(`brightness_min_${subSwitch}`) *
+      TUYA_PERCENTAGE_SCALING;
+    const scaleMax =
+      this.getSetting(`brightness_max_${subSwitch}`) *
+      TUYA_PERCENTAGE_SCALING;
```
Comment (bobvandevijver) [inline drivers/dimmer/device.js:213]:
```suggestion
    const scaleMin = this.getSetting(`brightness_min_${subSwitch}`) * TUYA_PERCENTAGE_SCALING;
    const scaleMax = this.getSetting(`brightness_max_${subSwitch}`) * TUYA_PERCENTAGE_SCALING;
```

File: drivers/dimmer/driver.js [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+        .getActionCard(`dimmer_channel_${switch_i}_dim`)
+        .registerRunListener(async (args) => {
+          await args.device.singleDim(args.value, `bright_value_${switch_i}`);
+        });
+    }
+  }
+
+  static TUYA_DEVICE_CATEGORIES = [
+    TuyaOAuth2Constants.DEVICE_CATEGORIES.LIGHTING.DIMMER,
+  ];
```
Comment (bobvandevijver) [inline drivers/dimmer/driver.js:40]: Let's put this on top

File: drivers/dimmer/device.js [hunk 8]
```diff
@@ -0,0 +1,223 @@
+"use strict";
+
+const TuyaOAuth2Device = require("../../lib/TuyaOAuth2Device");
+const { DIMMER_SETTING_LABELS } = require("./driver");
```
Comment (bobvandevijver) [inline drivers/dimmer/device.js:4]: Let's create a dedicated file for this

File: drivers/dimmer/driver.js [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+    ],
+  };
+
+  onTuyaPairListDeviceProperties(device, specification) {
+    const props = super.onTuyaPairListDeviceProperties(device);
+    props.store.tuya_switches = [];
+    props.store.tuya_dimmers = [];
+
+    const simpleCapabilities =
+      TuyaOAuth2DriverDimmer.SIMPLE_DIMMER_CAPABILITIES;
```
Comment (bobvandevijver) [inline drivers/dimmer/driver.js:77]:
```suggestion
    const simpleCapabilities = TuyaOAuth2DriverDimmer.SIMPLE_DIMMER_CAPABILITIES;
```

File: drivers/dimmer/driver.js [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+        props.settings["brightness_max_1"] =
+          values.max / TUYA_PERCENTAGE_SCALING;
+      }
+
+      if (tuyaCapability === "bright_value_2") {
+        props.settings["brightness_min_2"] =
+          values.min / TUYA_PERCENTAGE_SCALING;
+        props.settings["brightness_max_2"] =
+          values.max / TUYA_PERCENTAGE_SCALING;
+      }
```
Comment (bobvandevijver) [inline drivers/dimmer/driver.js:168]:
```suggestion
      if (tuyaCapability === "bright_value_1") {
        props.settings["brightness_min_1"] = values.min / TUYA_PERCENTAGE_SCALING;
        props.settings["brightness_max_1"] = values.max / TUYA_PERCENTAGE_SCALING;
      }

      if (tuyaCapability === "bright_value_2") {
        props.settings["brightness_min_2"] = values.min / TUYA_PERCENTAGE_SCALING;
        props.settings["brightness_max_2"] = values.max / TUYA_PERCENTAGE_SCALING;
      }
```

**Review by bobvandevijver — CHANGES_REQUESTED** [review]

File: drivers/dimmer/device.js (hunk 8 above)
Comment (JELoohuis) [reply drivers/dimmer/device.js]: Fixed in 6198a246d7072d6eeb71dad258c68a37ab5fa07b

**Review by bobvandevijver — APPROVED** [review]

**Pull request merged**